### PR TITLE
LibCore: Port AnonymousBuffer to Windows

### PR DESCRIPTION
**Pull request by stasoid**

### CODE REVIEW

File: Userland/Libraries/LibCore/AnonymousBufferWindows.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        CloseHandle(m_map_handle);
+}
+
+ErrorOr<NonnullRefPtr<AnonymousBufferImpl>> AnonymousBufferImpl::create(size_t size)
+{
+    static int next_id_tag;
+    ByteString name = ByteString::formatted("AB-{}-{}", GetCurrentProcessId(), next_id_tag++);
+
+    HANDLE map_handle = CreateFileMapping(INVALID_HANDLE_VALUE, NULL, PAGE_READWRITE,
+        HIWORD(size), LOWORD(size), name.characters());
```
Comment (ADKaster) [inline Userland/Libraries/LibCore/AnonymousBufferWindows.cpp]: This formatting looks off, it shouldn't be wrapped like this when using our clang-format-18 config. single-line is fine.

File: Userland/Libraries/LibCore/AnonymousBufferWindows.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        UnmapViewOfFile(m_data);
+
+    if (m_map_handle)
+        CloseHandle(m_map_handle);
+}
+
+ErrorOr<NonnullRefPtr<AnonymousBufferImpl>> AnonymousBufferImpl::create(size_t size)
+{
+    static int next_id_tag;
+    ByteString name = ByteString::formatted("AB-{}-{}", GetCurrentProcessId(), next_id_tag++);
```
Comment (ADKaster) [inline Userland/Libraries/LibCore/AnonymousBufferWindows.cpp]: Why do we need to name these? 

Per the docs https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createfilemappinga:

> [in, optional] lpName

> The name of the file mapping object.

> If this parameter matches the name of an existing mapping object, the function requests access to the object with the protection that flProtect specifies.

> If this parameter is NULL, the file mapping object is created without a name.

We shouldn't need to keep the name around. The only thing we should be pushing over IPC is a handle duplicated via `DuplicateHandle`, assuming those can be transferred over windows IPC mechanisms?

> Multiple processes can share a view of the same file by either using a single shared file mapping object or creating separate file mapping objects backed by the same file. A single file mapping object can be shared by multiple processes through inheriting the handle at process creation, duplicating the handle, or opening the file mapping object by name. For more information, see the [CreateProcess](https://learn.microsoft.com/en-us/windows/desktop/api/processthreadsapi/nf-processthreadsapi-createprocessa), [DuplicateHandle](https://learn.microsoft.com/en-us/windows/desktop/api/handleapi/nf-handleapi-duplicatehandle) and [OpenFileMapping](https://learn.microsoft.com/en-us/windows/desktop/api/winbase/nf-winbase-openfilemappinga) functions.

I would rather that we keep the "anonymous" behavior of Anonymous buffers as much as possible. It shouldn't be possible for other random processes on the system running as the current user to intercept and read our mapped file data by guessing the names.

File: Userland/Libraries/LibIPC/Decoder.cpp [hunk 3]
```diff
@@ -131,9 +131,14 @@ ErrorOr<Core::AnonymousBuffer> decode(Decoder& decoder)
...
     auto size = TRY(decoder.decode_size());
-    auto anon_file = TRY(decoder.decode<IPC::File>());
 
-    return Core::AnonymousBuffer::create_from_anon_fd(anon_file.take_fd(), size);
+#ifndef AK_OS_WINDOWS
+    auto anon_file = TRY(decoder.decode<IPC::File>());
+    auto id = anon_file.take_fd();
+#else
+    auto id = TRY(decoder.decode<ByteString>());
+#endif
```
Comment (ADKaster) [inline Userland/Libraries/LibIPC/Decoder.cpp:140]: here, I would rather update IPC::File to know whether it's holding a C/Unix `file descriptor` or a HANDLE and manage that internally, rather than having to add ifdefs for windows here in the decoder for anonymous buffer specifically.

File: Documentation/Patterns.md [hunk 4]
```diff
@@ -33,8 +33,8 @@ ErrorOr<NonnullRefPtr<Bitmap>> Bitmap::create_shareable(BitmapFormat format, Int
     auto const pitch = minimum_pitch(size.width() * scale_factor, format);
     auto const data_size = size_in_bytes(pitch, size.height() * scale_factor);
 
-    auto buffer = TRY(Core::AnonymousBuffer::create_with_size(round_up_to_power_of_two(data_size, PAGE_SIZE)));
-    auto bitmap = TRY(Bitmap::create_with_anonymous_buffer(format, buffer, size, scale_factor, {}));
```
Comment (ADKaster) [inline Documentation/Patterns.md:37]: I'm not sure I understand the motivation behind renaming Anonymous buffer. One of the main features of anonymous buffer is that it models the "memfd" concept in linux, where the only actors that have access to the data are ones with whom a handle has been shared specifically.

**Review by ADKaster — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibCore/AnonymousBufferWindows.cpp (hunk 2 above)
Comment (stasoid) [reply Userland/Libraries/LibCore/AnonymousBufferWindows.cpp]: Because using names is the only way to keep current architecture. `DuplicateHandle` requres target process pid. So IPC::Encoder would require to know target process pid.

I think we should worry only about inadvertent clash of names with other processes. If someone is malicious and tries to guess names they might also just read/write process memory directly with ReadProcessMemory.

File: Documentation/Patterns.md (hunk 4 above)
Comment (stasoid) [reply Documentation/Patterns.md:37]: When I first saw this name, I was confused: aren't all buffers anonymous? Surely all buffers I encountered before were anonymous.
Yes, one of the features is that it models memfd. But the main feature, its primary purpose is shared buffer. memfd is just an implementation detail.
And by the way, even not considering Windows implementaion, anon_create on some systems (BSD generic, Haiku) is implemented using named shm_open.

To summarize: the name AnonymousBuffer is confusing, doesn't reflect its primary purpose and just plain wrong on some platforms.

File: Userland/Libraries/LibCore/AnonymousBufferWindows.cpp (hunk 2 above)
Comment (ADKaster) [reply Userland/Libraries/LibCore/AnonymousBufferWindows.cpp]: Does it not work to pass DUPLICATE_CLOSE_SOURCE in the options and null for hTargetProcessHandle? If we are ok with making a bunch of extra kernel objects, we could even first duplicate the handle passing ourselves, then duplicate the initial handle with "close source" to get an anonymous handle that can be passed to anyone. Assuming the API works the way I think it does, anyway.

We have a similar constraint for mach ports that we need the parent process to receive a send port right from the child in order to send it data over IPC. I recently added a new abstraction to LibIPC called "transport", which should help abstract away different ipc mechanisms between OS platforms.

File: Userland/Libraries/LibIPC/Decoder.cpp (hunk 3 above)
Comment (stasoid) [reply Userland/Libraries/LibIPC/Decoder.cpp:140]: AnonymousBuffer on Windows doesn't use IPC::File, it just passes its name.

File: Userland/Libraries/LibCore/AnonymousBufferWindows.cpp (hunk 2 above)
Comment (stasoid) [reply Userland/Libraries/LibCore/AnonymousBufferWindows.cpp]: No, this doesn't work:
>[out] lpTargetHandle
...
This parameter is ignored if hTargetProcessHandle is NULL.

Comment (stasoid) [reply Userland/Libraries/LibCore/AnonymousBufferWindows.cpp]: I am currently working on EventLoop, and it has similar situation. When choosing how to implement Unix unnamed pipes and Unix domain sockets I chose Windows named pipes over sockets because when IPC::encode()ing a socket handle we need to know target process id for WSADuplicateSocket. (I don't use Windows unnamed pipes because they are not waitable)

I encourage you not to be very picky about implementation details at this point. I think when we have no functionality at all anything that works will do. If this architecture is deemed inappropriate it can always be upgraded later.

File: Documentation/Patterns.md (hunk 4 above)
Comment (stasoid) [reply Documentation/Patterns.md]: No response => failed to convince => no renaming

**Comment by stasoid** [comment]
I use m_fd on Windows to store file mapping handle. I know this is undesirable, but the alternative is much much uglier - AnonymousBuffer.h becomes riddled with OS-specific #ifdefs, and it cannot be solved just by defining types, because the concepts itself have different names (file descriptor vs handle):
```cpp
/*
 * Copyright (c) 2021, Andreas Kling <andreas@ladybird.org>
 * Copyright (c) 2022, the SerenityOS developers.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */

#pragma once

#include <AK/Error.h>
#include <AK/Noncopyable.h>
#include <AK/RefCounted.h>
#include <AK/RefPtr.h>
#include <AK/Types.h>
#include <LibIPC/Forward.h>

namespace Core {

class AnonymousBufferImpl final : public RefCounted<AnonymousBufferImpl> {
public:
#ifndef AK_OS_WINDOWS
    static ErrorOr<NonnullRefPtr<AnonymousBufferImpl>> create(int fd, size_t);
#else
    static ErrorOr<NonnullRefPtr<AnonymousBufferImpl>> create(size_t);
    static ErrorOr<NonnullRefPtr<AnonymousBufferImpl>> create(void* handle, size_t);
#endif
    ~AnonymousBufferImpl();

#ifndef AK_OS_WINDOWS
    int fd() const { return m_fd; }
#else
    void* handle() const { return m_handle; }
#endif
    size_t size() const { return m_size; }
    void* data() { return m_data; }
    void const* data() const { return m_data; }

private:
#ifndef AK_OS_WINDOWS
    AnonymousBufferImpl(int fd, size_t, void*);

    int m_fd { -1 };
#else
    AnonymousBufferImpl(void* handle, size_t, void*);

    void* m_handle { nullptr };
#endif
    size_t m_size { 0 };
    void* m_data { nullptr };
};

class AnonymousBuffer {
public:
    static ErrorOr<AnonymousBuffer> create_with_size(size_t);
#ifndef AK_OS_WINDOWS
    static ErrorOr<AnonymousBuffer> create_from_anon_fd(int fd, size_t);
#else
    static ErrorOr<AnonymousBuffer> create_from_handle(void* handle, size_t);
#endif

    AnonymousBuffer() = default;

    bool is_valid() const { return m_impl; }

#ifndef AK_OS_WINDOWS
    int fd() const { return m_impl ? m_impl->fd() : -1; }
#else
    int handle() const { return m_impl ? m_impl->handle() : nullptr; }
#endif
    size_t size() const { return m_impl ? m_impl->size() : 0; }

    template<typename T>
    T* data()
    {
        static_assert(IsVoid<T> || IsTrivial<T>);
        if (!m_impl)
            return nullptr;
        return (T*)m_impl->data();
    }

    template<typename T>
    T const* data() const
    {
        static_assert(IsVoid<T> || IsTrivial<T>);
        if (!m_impl)
            return nullptr;
        return (T const*)m_impl->data();
    }

private:
    explicit AnonymousBuffer(NonnullRefPtr<AnonymousBufferImpl> impl)
        : m_impl(move(impl))
    {
    }

    RefPtr<AnonymousBufferImpl> m_impl;
};

}

namespace IPC {

template<>
ErrorOr<void> encode(Encoder&, Core::AnonymousBuffer const&);

template<>
ErrorOr<Core::AnonymousBuffer> decode(Decoder&);

}
```
The only viable alternative I see is to use "handle" on all OSes, because it is a more general concept than fd, but then Linux programmers would be confused: why are we calling a file descriptor "handle"?

File: Userland/Libraries/LibCore/AnonymousBufferWindows.cpp (hunk 2 above)
Comment (stasoid) [reply Userland/Libraries/LibCore/AnonymousBufferWindows.cpp]: Updated to use handle instead of name.

**Comment by stasoid** [comment]
Or maybe invent some other concept that maps to fd on Linux and to handle on Windows?
Like "id"? m_id is not much different from m_fd, which is good.

**Review by ADKaster — APPROVED** [review]
Thanks, this implementation looks a lot simpler. There's a future where we could use `FlatPtr` or similar concepts to abstract things away in a slightly cleaner fashion than plain int.
